File: solutions/ruby/0712-minimum-ascii-delete-sum-for-two-strings.rb
# @param {String} s1
# @param {String} s2
# @return {Integer}
def minimum_delete_sum(s1, s2)
    prev_row = Array.new(s2.length + 1, 0)
    (1..s2.length).each { |j| prev_row[j] = prev_row[j - 1] + s2[j - 1].ord }

    (1..s1.length).each do |i|
        curr_row = [prev_row[0] + s1[i - 1].ord]
        (1..s2.length).each do |j|
            if s1[i - 1] == s2[j - 1]
                curr_row << prev_row[j - 1]
            else
                curr_row << [prev_row[j] + s1[i - 1].ord, curr_row[j - 1] + s2[j - 1].ord].min
            end
        end
        prev_row = curr_row
    end

    prev_row[-1]
end

File: solutions/javascript/0712-minimum-ascii-delete-sum-for-two-strings.js
/**
 * @param {string} s1
 * @param {string} s2
 * @return {number}
 */
var minimumDeleteSum = function(s1, s2) {
    let prev_row = new Array(s2.length + 1).fill(0);
        for (let j = 1; j <= s2.length; j++)
            prev_row[j] = prev_row[j - 1] + s2.charCodeAt(j - 1);
        
        for (let i = 1; i <= s1.length; i++) {
            let curr_row = [prev_row[0] + s1.charCodeAt(i - 1)];
            for (let j = 1; j <= s2.length; j++) {
                if (s1[i - 1] === s2[j - 1])
                    curr_row.push(prev_row[j - 1]);
                else
                    curr_row.push(Math.min(prev_row[j] + s1.charCodeAt(i - 1), curr_row[curr_row.length - 1] + s2.charCodeAt(j - 1)));
            }
            prev_row = curr_row;
        }
        
        return prev_row[prev_row.length - 1];
};
